server/src/main/java/dev/slimevr/tracking/processor/skeleton/Localizer.kt

```diff
@@ -0,0 +1,425 @@
+		return loc.minus(targetFoot)
+	}
+
+	// update the target position of the foot
+	private fun updateTargetPos(loc: Vector3, foot: MovementStates) {
+		if (foot == plantedFoot) {
+			if (worldReference === MovementStates.FOLLOW_COM) {
+				targetFoot = loc
+			}
+		} else {
+			targetFoot = loc
+			plantedFoot = foot
+		}
+	}
+
+	// get the sitting travel (emulates hip lock)
+	private fun getSittingTravel(): Vector3 {
+		val hip: Vector3 = skeleton.computedHipTracker.position
+
+		// get the distance to move the waist to the target waist
+		val dist: Vector3 = hip.minus(targetHip)
+
+		// if the world reference is not sitting update the target waist
+		if (worldReference != MovementStates.FOLLOW_SITTING || sittingFrames < SITTING_EARLY) {
+			targetHip = hip
+		}
+		return dist
+	}
+
+	// returns the travel of the COM from its last position
+	private fun getCOMTravel(): Vector3 {
+		// update COM attributes
+		updateCOMAttributes()
+
+		var dist: Vector3 = bufCur.centerOfMass
+		dist = dist.minus(targetCOM)
+
+		return dist
+	}
+
+	// get the movement of the COM based on the last velocity
+	private fun updateCOMAttributes() {
+		getCOMVelocity()
+		updateTargetCOM()
+
+		// update how long the COM has been the reference and how long the foot
+		// has been
+		comFrames = if (worldReference === MovementStates.FOLLOW_COM) comFrames + 1 else 0
+		footFrames = if (worldReference === MovementStates.FOLLOW_FOOT) footFrames + 1 else 0
+		sittingFrames = if (worldReference === MovementStates.FOLLOW_SITTING) sittingFrames + 1 else 0
+	}
+
+	// gets the position the COM should be at based on the velocity of the com and
+	// the location of the floor
+	private fun updateTargetCOM() {
+		// if not in COM tracking mode, just use the current COM
+		if (worldReference === MovementStates.FOLLOW_FOOT || worldReference === MovementStates.FOLLOW_SITTING) {
+			targetCOM = bufCur.centerOfMass
+		} else {
+			currentCOM = targetCOM
+		}
+
+		targetCOM = targetCOM.plus(comVelocity.div(bufCur.timeDelta))
+
+		val lowTracker = getLowestTracker()
+
+		// update the target COM and velocity to reflect this new distance
+		if (lowTracker.position.y < uncorrectedFloor) {
+			targetCOM = Vector3(targetCOM.x, targetCOM.y + (uncorrectedFloor - lowTracker.position.y), targetCOM.z)
+			comVelocity = Vector3(comVelocity.x, 0.0f, comVelocity.z)
+		}
+	}
+
+	// get the velocity of the COM
+	private fun getCOMVelocity(): Vector3 {
+		val comY = comVelocity.y
+
+		var buf = bufCur
+		val timeStart: Long = buf.timeOfFrame
+		var timeEnd = timeStart - VELOCITY_SAMPLE_RATE
+		val comPosStart: Vector3 = buf.centerOfMass
+
+		// get the buffer that occurred VELOCITY_SAMPLE_RATE ago in time
+		while (buf.timeOfFrame > timeEnd && buf.parent !== null) {
+			buf = buf.parent
+		}
+
+		val comPosEnd: Vector3 = buf.centerOfMass
+		timeEnd = buf.timeOfFrame
+
+		// calculate the velocity
+		comVelocity = comPosStart.minus(comPosEnd).div((timeStart - timeEnd) / LegTweakBuffer.NS_CONVERT)
```

Comment:
Shouldn't this be reversed? I feel like I'm missing something here lol, it looks like it'd technically work anyways though because of negation
```suggestion
		comVelocity = comPosEnd.minus(comPosStart).div((timeEnd - timeStart) / LegTweakBuffer.NS_CONVERT)
```

Reply:
I got super lucky and it evaluates to the same thing, It should be flipped though for readability